### PR TITLE
Add: Kubernetes recommande labels

### PR DESCRIPTION
**Pull request by nlamirault**
Signed-off-by: Nicolas Lamirault <nicolas.lamirault@gmail.com>

**What this PR does / why we need it**:

Add Kubernetes recommanded labels https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

Ex:

```yaml
---
# Source: kube-state-metrics/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app: kube-state-metrics
    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
  name: RELEASE-NAME-kube-state-metrics
  namespace: default
imagePullSecrets:
  []
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

Output : 

```
$ helm template charts/kube-state-metrics
```

```yaml
---
# Source: kube-state-metrics/templates/serviceaccount.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app: kube-state-metrics    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
  name: RELEASE-NAME-kube-state-metrics
  namespace: default
imagePullSecrets:
  []
---
# Source: kube-state-metrics/templates/role.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    app: kube-state-metrics    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
  name: RELEASE-NAME-kube-state-metrics
rules:

- apiGroups: ["certificates.k8s.io"]
  resources:
  - certificatesigningrequests
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - configmaps
  verbs: ["list", "watch"]

- apiGroups: ["batch"]
  resources:
  - cronjobs
  verbs: ["list", "watch"]

- apiGroups: ["extensions", "apps"]
  resources:
  - daemonsets
  verbs: ["list", "watch"]

- apiGroups: ["extensions", "apps"]
  resources:
  - deployments
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - endpoints
  verbs: ["list", "watch"]

- apiGroups: ["autoscaling"]
  resources:
  - horizontalpodautoscalers
  verbs: ["list", "watch"]

- apiGroups: ["extensions", "networking.k8s.io"]
  resources:
  - ingresses
  verbs: ["list", "watch"]

- apiGroups: ["batch"]
  resources:
  - jobs
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - limitranges
  verbs: ["list", "watch"]

- apiGroups: ["admissionregistration.k8s.io"]
  resources:
    - mutatingwebhookconfigurations
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - namespaces
  verbs: ["list", "watch"]

- apiGroups: ["networking.k8s.io"]
  resources:
  - networkpolicies
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - nodes
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - persistentvolumeclaims
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - persistentvolumes
  verbs: ["list", "watch"]

- apiGroups: ["policy"]
  resources:
    - poddisruptionbudgets
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - pods
  verbs: ["list", "watch"]

- apiGroups: ["extensions", "apps"]
  resources:
  - replicasets
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - replicationcontrollers
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - resourcequotas
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - secrets
  verbs: ["list", "watch"]

- apiGroups: [""]
  resources:
  - services
  verbs: ["list", "watch"]

- apiGroups: ["apps"]
  resources:
  - statefulsets
  verbs: ["list", "watch"]

- apiGroups: ["storage.k8s.io"]
  resources:
    - storageclasses
  verbs: ["list", "watch"]

- apiGroups: ["admissionregistration.k8s.io"]
  resources:
    - validatingwebhookconfigurations
  verbs: ["list", "watch"]

- apiGroups: ["storage.k8s.io"]
  resources:
    - volumeattachments
  verbs: ["list", "watch"]
---
# Source: kube-state-metrics/templates/clusterrolebinding.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  labels:
    app: kube-state-metrics    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
  name: RELEASE-NAME-kube-state-metrics
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: RELEASE-NAME-kube-state-metrics
subjects:
- kind: ServiceAccount
  name: RELEASE-NAME-kube-state-metrics
  namespace: default
---
# Source: kube-state-metrics/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: RELEASE-NAME-kube-state-metrics
  namespace: default
  labels:
    app: kube-state-metrics    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
  annotations:
    prometheus.io/scrape: 'true'
spec:
  type: "ClusterIP"
  ports:
  - name: "http"
    protocol: TCP
    port: 8080
    targetPort: 8080
  
  selector:
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
---
# Source: kube-state-metrics/templates/deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kube-state-metrics
  namespace: default
  labels:
    app: kube-state-metrics    
    app.kubernetes.io/name: kube-state-metrics
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: "2.13.0"
    app.kubernetes.io/part-of: kube-state-metrics 
    helm.sh/chart: kube-state-metrics-2.13.0
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: kube-state-metrics
  replicas: 1
  template:
    metadata:
      labels:
        app: kube-state-metrics        
        app.kubernetes.io/name: kube-state-metrics
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/version: "2.13.0"
        app.kubernetes.io/part-of: kube-state-metrics 
        helm.sh/chart: kube-state-metrics-2.13.0
    spec:
      hostNetwork: false
      serviceAccountName: RELEASE-NAME-kube-state-metrics
      securityContext:
        fsGroup: 65534
        runAsGroup: 65534
        runAsUser: 65534
      containers:
      - name: kube-state-metrics
        args:


        - --collectors=certificatesigningrequests


        - --collectors=configmaps


        - --collectors=cronjobs


        - --collectors=daemonsets


        - --collectors=deployments


        - --collectors=endpoints


        - --collectors=horizontalpodautoscalers


        - --collectors=ingresses


        - --collectors=jobs


        - --collectors=limitranges


        - --collectors=mutatingwebhookconfigurations


        - --collectors=namespaces


        - --collectors=networkpolicies


        - --collectors=nodes


        - --collectors=persistentvolumeclaims


        - --collectors=persistentvolumes


        - --collectors=poddisruptionbudgets


        - --collectors=pods


        - --collectors=replicasets


        - --collectors=replicationcontrollers


        - --collectors=resourcequotas


        - --collectors=secrets


        - --collectors=services


        - --collectors=statefulsets


        - --collectors=storageclasses


        - --collectors=validatingwebhookconfigurations



        - --collectors=volumeattachments





        - --telemetry-port=8081
        imagePullPolicy: IfNotPresent
        image: "quay.io/coreos/kube-state-metrics:v1.9.7"
        ports:
        - containerPort: 8080
        livenessProbe:
          httpGet:
            path: /healthz
            port: 8080
          initialDelaySeconds: 5
          timeoutSeconds: 5
        readinessProbe:
          httpGet:
            path: /
            port: 8080
          initialDelaySeconds: 5
          timeoutSeconds: 5
```

### CODE REVIEW

**Comment by mrueg** [comment]
Thanks for your contribution!
To make sure I understood the PR correctly, you don't add the labels and want to move them into a template macro?

File: charts/kube-state-metrics/templates/_helpers.tpl [hunk 1]
```diff
@@ -24,6 +24,19 @@ If release name contains chart name it will be used as a full name.
...
+{{/* Generate basic labels */}}
+{{- define "kube-state-metrics.labels" }}
+app.kubernetes.io/name: {{ template "kube-state-metrics.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+app.kubernetes.io/managed-by: {{ .Release.Service }}
+app.kubernetes.io/version: "{{ .Chart.Version }}"
+app.kubernetes.io/part-of: {{ template "kube-state-metrics.name" . }} 
+helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version }} 
+{{- if .Values.customLabels }}
+{{ toYaml .Values.customLabels | indent 4 }}
```
Comment (mrueg) [inline charts/kube-state-metrics/templates/_helpers.tpl:36]: Is the indent here working correctly?

Comment (nlamirault) [reply charts/kube-state-metrics/templates/_helpers.tpl:36]: i make an output :

```
$ helm template charts/kube-state-metrics > output.yaml
```

File: charts/kube-state-metrics/templates/clusterrolebinding.yaml [hunk 2]
```diff
@@ -3,10 +3,8 @@ apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRoleBinding
 metadata:
   labels:
-    app.kubernetes.io/name: {{ template "kube-state-metrics.name" . }}
-    helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version }}
-    app.kubernetes.io/managed-by: {{ .Release.Service }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
+    app: {{ template "kube-state-metrics.name" . }}
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply charts/kube-state-metrics/templates/clusterrolebinding.yaml:6]: I do that to be consistent with other charts (Like : https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/prometheus/clusterrolebinding.yaml#L7).
Do you want i remove it and use customLabels ?

**Comment by mrueg** [comment]
/hold
Thanks for a contribution!
We are currently migrating the chart to a new location. Please feel free to create a new PR in https://github.com/prometheus-community/helm-charts/ once the migration has been completed.

https://github.com/kubernetes/kube-state-metrics/pull/1448

**Comment by mrueg** [comment]
Thanks for your contribution. We moved our helm charts to https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics Feel free to reopen a PR on https://github.com/prometheus-community/helm-charts/ !

**Pull request closed without merging**